File: ruby/day7_part2_2025.rb
#!/usr/bin/env ruby
file = 'input.txt'
unless File.exist?(file)
  $stderr.puts 'File not found'
  exit 1
end
grid = File.readlines(file).map(&:chomp).reject(&:empty?)
if grid.empty?
  puts 0
  exit
end

h = grid.size
w = grid[0].length
sx = sy = nil
grid.each_with_index do |row, y|
  x = row.index('S')
  if x
    sx, sy = x, y
    break
  end
end
unless sx
  $stderr.puts "Start point 'S' not found"
  exit 1
end

counts = { sx => 1 }
(sy...h).each do |y|
  nxt = Hash.new(0)
  counts.each do |x, cnt|
    if x >= 0 && x < w && grid[y][x] == '^'
      nxt[x - 1] += cnt
      nxt[x + 1] += cnt
    else
      nxt[x] += cnt
    end
  end
  counts = nxt
end

total = counts.values.reduce(0, :+)
puts total

File: julia/day7_part2_2025.jl

#!/usr/bin/env julia
using Base: @inbounds

function add!(d::Dict{Int,BigInt}, k::Int, v::BigInt)
    d[k] = get(d, k, big(0)) + v
end

function main()
    # read non‑empty lines
    lines = filter(!isempty, readlines("input.txt"))
    if isempty(lines)
        println(0)
        return
    end

    h = length(lines)                     # number of rows
    w = length(first(lines))              # number of columns (all rows have same length)

    # locate start point 'S' (zero‑based coordinates)
    sx = -1; sy = -1
    for y in 0:h-1
        row = lines[y+1]
        for x in 0:w-1
            @inbounds if row[x+1] == 'S'
                sx = x; sy = y
                break
            end
        end
        sx != -1 && break
    end
    sx == -1 && error("Start point 'S' not found")

    cur = Dict{Int,BigInt}(sx => big(1))

    for y in sy:h-1                     # process from the start row downwards
        row = lines[y+1]                # convert to 1‑based indexing for the string
        nxt = Dict{Int,BigInt}()

        for (x, cnt) in cur
            if 0 ≤ x < w && row[x+1] == '^'
                add!(nxt, x-1, cnt)
                add!(nxt, x+1, cnt)
            else
                add!(nxt, x, cnt)
            end
        end
        cur = nxt
    end

    total = sum(values(cur))
    println(string(total))
end

main()
